src/hotspot/share/nmt/nmtCommon.hpp

```diff
@@ -137,5 +141,32 @@ class NMTUtil : AllStatic {
   static S _strings[mt_number_of_types];
 };
 
+class NmtGuard : public StackObj {
+private:
+    static Semaphore _nmt_semaphore;
+    static intx volatile _owner;
+
+public:
+    NmtGuard() {
+      intx const current =  os::current_thread_id();
+      intx const owner = Atomic::load(&_owner);
+      assert(current != owner, "Lock is not reentrant");
+
+      _nmt_semaphore.wait();
+      Atomic::store(&_owner, current);
+    }
+
+    ~NmtGuard() {
+      assert_locked();
+      Atomic::store(&_owner, (intx) -1);
+      _nmt_semaphore.signal();
+    }
+
+    static void assert_locked(){
+      intx const current = os::current_thread_id();
+      intx const owner = Atomic::load(&_owner);
+      assert(current == owner, "NMT lock should be acquired in this section.");
```

Comment:
How about we try to re-use `assert_locked()` in constructor and destructor, something like:

```
    NmtGuard() {
      assert(!assert_locked(), "Lock is not reentrant");
      _nmt_semaphore.wait();
      Atomic::store(&_owner, current);
    }

    ~NmtGuard() {
      assert(assert_locked(), "NMT lock should be acquired in this section.");
      Atomic::store(&_owner, (intx) -1);
      _nmt_semaphore.signal();
    }

    static bool assert_locked(){
      intx const current = os::current_thread_id();
      intx const owner = Atomic::load(&_owner);
      return (current == owner);
    }

```

Reply:
I think it's a good idea to use it in the destructor. But based on the discussion above, it seems like we shouldn't assert the lock can't be reentered.